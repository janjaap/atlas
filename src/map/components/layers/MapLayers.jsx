import React from 'react';
import PropTypes from 'prop-types';

import MapLegend from '../legend/MapLegend';
import AddIcon from '../../../../public/images/icon-plus.svg';
import RemoveIcon from '../../../../public/images/icon-cross.svg';

import './_map-layers.scss';

const getCategories = (layers) => (
  layers.filter((layer) => layer.category).map((layer) => layer.category)
);

const MapLayers = (props) => (
  <div className="map-layers">
    <h3 className="u-sr-only">Beschikbare kaartlagen</h3>
    <ul>
      {[...new Set(getCategories(props.layers))]
        .map((category) => (
          <li className="map-layers__category" key={category}>
            <h4 className="map-layers__category-text">{category}</h4>
            <ul>
              {props.layers
                .filter((layer) => layer.category === category)
                .map((layer) => (
                  <li
                    className={`
                      map-layers__title
                      map-layers__title--${props.activeMapLayers.some((mapLayer) => layer.title === mapLayer.title) ? 'active' : 'inactive'}
                    `}
                    key={layer.title}
                  >
                    <button onClick={() => {
                      MapLegend.mapLayersLegendItemsToIds(layer).forEach((mapLayerId) =>
                        props.onLayerToggle(mapLayerId));
                    }}
                    >
                      <span className="map-layers__toggle-title">
                        {layer.title}
                      </span>
                      <span className="map-layers__toggle map-layers__toggle--remove">
                        <RemoveIcon />
                      </span>
                      <span className="map-layers__toggle map-layers__toggle--add">
                        <AddIcon />
                      </span>
                    </button>
                  </li>
              ))}
            </ul>
          </li>
      ))}
    </ul>
  </div>
);

MapLayers.propTypes = {
  activeMapLayers: PropTypes.array, // eslint-disable-line
  layers: PropTypes.array, // eslint-disable-line
  onLayerToggle: PropTypes.func.isRequired // eslint-disable-line react/no-unused-prop-types
};

export default MapLayers;
