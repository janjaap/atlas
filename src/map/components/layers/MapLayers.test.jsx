import React from 'react';
import { shallow } from 'enzyme';
import MapLayers from './MapLayers';

const layers = [
  {
    id: 'pano2018',
    layers: [
      'panorama_recent_2018'
    ],
    legendItems: [
      {
        selectable: false,
        title: '2018'
      }
    ],
    maxZoom: 16,
    minZoom: 11,
    title: 'Panoramabeelden',
    url: '/maps/panorama?version=1.3.0&service=WMS'
  },
  {
    id: 'pano2017',
    layers: [
      'panorama_recent_2017'
    ],
    legendItems: [
      {
        selectable: false,
        title: '2017'
      }
    ],
    maxZoom: 16,
    minZoom: 11,
    title: 'Panoramabeelden',
    url: '/maps/panorama?version=1.3.0&service=WMS'
  },
  {
    id: 'pano2016',
    layers: [
      'panorama_recent_2016'
    ],
    legendItems: [
      {
        selectable: false,
        title: '2016'
      }
    ],
    maxZoom: 16,
    minZoom: 11,
    title: 'Panoramabeelden',
    url: '/maps/panorama?version=1.3.0&service=WMS'
  },
  {
    category: 'Geografie: onroerende zaken',
    legendItems: [
      {
        id: 'bgem',
        notClickable: true,
        layer: 'burgerlijke_gemeente',
        selectable: true,
        title: 'Burgerlijke gemeente'
      },
      {
        id: 'kgem',
        notClickable: true,
        layer: 'kadastrale_gemeente',
        selectable: true,
        title: 'Kadastrale gemeente'
      },
      {
        id: 'ksec',
        notClickable: true,
        layer: 'kadastrale_sectie',
        selectable: true,
        title: 'Kadastrale sectie'
      },
      {
        id: 'kot',
        layer: 'kadastraal_object',
        selectable: true,
        title: 'Kadastraal object'
      }
    ],
    maxZoom: 16,
    minZoom: 8,
    title: 'Kadastrale perceelsgrenzen',
    url: '/maps/brk?version=1.3.0&service=WMS'
  }
];

const activeLayers = [
  {
    category: 'Geografie: onroerende zaken',
    legendItems: [
      {
        id: 'bgem',
        notClickable: true,
        layer: 'burgerlijke_gemeente',
        selectable: true,
        title: 'Burgerlijke gemeente'
      },
      {
        id: 'kgem',
        notClickable: true,
        layer: 'kadastrale_gemeente',
        selectable: true,
        title: 'Kadastrale gemeente'
      },
      {
        id: 'ksec',
        notClickable: true,
        layer: 'kadastrale_sectie',
        selectable: true,
        title: 'Kadastrale sectie 1'
      }
    ],
    maxZoom: 16,
    minZoom: 8,
    title: 'Kadastrale perceelsgrenzen',
    url: '/maps/brk?version=1.3.0&service=WMS'
  }
];

describe('MapLayer', () => {
  it('should render', () => {
    const component = shallow(
      <MapLayers
        activeMapLayers={activeLayers}
        layers={layers}
        onLayerToggle={() => {
        }}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render without active layers', () => {
    const component = shallow(
      <MapLayers
        activeMapLayers={[]}
        layers={layers}
        onLayerToggle={() => {
        }}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should call onLayerToggle 4 times when user clicks on button', () => {
    const onLayerToggleMock = jest.fn();
    const component = shallow(
      <MapLayers
        activeMapLayers={activeLayers}
        layers={layers}
        onLayerToggle={onLayerToggleMock}
      />
    );

    const button = component.find('button');
    button.simulate('click');

    expect(onLayerToggleMock).toHaveBeenCalledTimes(4);
  });
});
