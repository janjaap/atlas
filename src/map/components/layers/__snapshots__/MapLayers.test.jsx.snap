// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapLayer should render 1`] = `
<div
  className="map-layers"
>
  <h3
    className="u-sr-only"
  >
    Beschikbare kaartlagen
  </h3>
  <ul>
    <li
      className="map-layers__category"
      key="Geografie: onroerende zaken"
    >
      <h4
        className="map-layers__category-text"
      >
        Geografie: onroerende zaken
      </h4>
      <ul>
        <li
          className="
                      map-layers__title
                      map-layers__title--active
                    "
          key="Kadastrale perceelsgrenzen"
        >
          <button
            onClick={[Function]}
          >
            <span
              className="map-layers__toggle-title"
            >
              Kadastrale perceelsgrenzen
            </span>
            <span
              className="map-layers__toggle map-layers__toggle--remove"
            />
            <span
              className="map-layers__toggle map-layers__toggle--add"
            />
          </button>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`MapLayer should render without active layers 1`] = `
<div
  className="map-layers"
>
  <h3
    className="u-sr-only"
  >
    Beschikbare kaartlagen
  </h3>
  <ul>
    <li
      className="map-layers__category"
      key="Geografie: onroerende zaken"
    >
      <h4
        className="map-layers__category-text"
      >
        Geografie: onroerende zaken
      </h4>
      <ul>
        <li
          className="
                      map-layers__title
                      map-layers__title--inactive
                    "
          key="Kadastrale perceelsgrenzen"
        >
          <button
            onClick={[Function]}
          >
            <span
              className="map-layers__toggle-title"
            >
              Kadastrale perceelsgrenzen
            </span>
            <span
              className="map-layers__toggle map-layers__toggle--remove"
            />
            <span
              className="map-layers__toggle map-layers__toggle--add"
            />
          </button>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
